File: app/src/main/java/com/meiji/daily/mvp/base/BaseFragment.java
package com.meiji.daily.mvp.base;

import android.os.Bundle;
import android.support.annotation.Nullable;
import android.support.v7.widget.Toolbar;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import com.trello.rxlifecycle2.LifecycleTransformer;
import com.trello.rxlifecycle2.android.FragmentEvent;
import com.trello.rxlifecycle2.components.support.RxFragment;

import javax.inject.Inject;

import me.drakeet.multitype.MultiTypeAdapter;

/**
 * Created by Meiji on 2017/5/5.
 */

public abstract class BaseFragment<T extends IBasePresenter> extends RxFragment implements IBaseView<T> {

    @Inject
    protected T presenter;
    protected MultiTypeAdapter adapter;

    /**
     * 绑定布局文件
     *
     * @return 布局文件ID
     */
    protected abstract int attachLayoutId();

    /**
     * 初始化视图控件
     */
    protected abstract void initViews(View view);

    /**
     * 初始化数据
     */
    protected abstract void initData();

    /**
     * Dagger 注入
     */
    protected abstract void initInjector();

    /**
     * 初始化 Toolbar
     *
     * @param toolbar
     * @param homeAsUpEnabled
     * @param title
     */
    protected void initToolBar(Toolbar toolbar, boolean homeAsUpEnabled, String title) {
        ((BaseActivity) getActivity()).initToolBar(toolbar, homeAsUpEnabled, title);
    }

    @Override
    public void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        initInjector();
    }

    @Nullable
    @Override
    public View onCreateView(LayoutInflater inflater, @Nullable ViewGroup container, @Nullable Bundle savedInstanceState) {
        View view = inflater.inflate(attachLayoutId(), container, false);
        initViews(view);
        initData();
        return view;
    }

    /**
     * 绑定生命周期
     */
    @Override
    public <T> LifecycleTransformer<T> bindToLife() {
        return bindUntilEvent(FragmentEvent.DESTROY);
    }
}

File: app/src/main/java/com/meiji/daily/module/base/BaseFragment.kt
package com.meiji.daily.module.base

import android.os.Bundle
import android.support.v4.app.Fragment
import android.support.v7.widget.Toolbar
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup

/**
 * Created by Meiji on 2017/12/4.
 */

abstract class BaseFragment : Fragment() {

    /**
     * 绑定布局文件
     *
     * @return 布局文件ID
     */
    protected abstract fun attachLayoutId(): Int

    /**
     * 初始化视图控件
     */
    protected abstract fun initViews(view: View)

    /**
     * 初始化数据
     */
    protected fun initData() {}

    /**
     * 订阅UI组件
     */
    protected open fun subscribeUI() {}

    /**
     * 初始化 Dagger
     */
    protected open fun initInject() {}

    /**
     * 初始化 Toolbar
     *
     * @param toolbar
     * @param homeAsUpEnabled
     * @param title
     */
    protected fun initToolBar(toolbar: Toolbar, homeAsUpEnabled: Boolean, title: String) {
        if (activity != null) {
            (activity as BaseActivity).initToolBar(toolbar, homeAsUpEnabled, title)
        }
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        initInject()
    }

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View? {
        val view = inflater.inflate(attachLayoutId(), container, false)
        initViews(view)
        initData()
        return view
    }

    override fun onActivityCreated(savedInstanceState: Bundle?) {
        super.onActivityCreated(savedInstanceState)
        if (!isAdded) {
            return
        }
        subscribeUI()
    }
}
